File: java/src/leetcode/editor/cn/P718_MaximumLengthOfRepeatedSubarray.java
//给两个整数数组 nums1 和 nums2 ，返回 两个数组中 公共的 、长度最长的子数组的长度 。 
//
// 
//
// 示例 1： 
//
// 
//输入：nums1 = [1,2,3,2,1], nums2 = [3,2,1,4,7]
//输出：3
//解释：长度最长的公共子数组是 [3,2,1] 。
// 
//
// 示例 2： 
//
// 
//输入：nums1 = [0,0,0,0,0], nums2 = [0,0,0,0,0]
//输出：5
// 
//
// 
//
// 提示： 
//
// 
// 1 <= nums1.length, nums2.length <= 1000 
// 0 <= nums1[i], nums2[i] <= 100 
// 
//
// Related Topics 数组 二分查找 动态规划 滑动窗口 哈希函数 滚动哈希 👍 790 👎 0

package leetcode.editor.cn.java;

//Java:最长重复子数组
//Time:2022-09-17 16:59:42
class P718_MaximumLengthOfRepeatedSubarray {
    public static void main(String[] args) {
        Solution solution = new P718_MaximumLengthOfRepeatedSubarray().new Solution();
//        System.out.println(solution.findLength(new int[]{1, 2, 3, 2, 1}, new int[]{3, 2, 1, 4, 7}));
        System.out.println(solution.findLength(new int[]{1, 0, 0, 0, 1}, new int[]{1, 0, 0, 1, 1}));
    }

    //leetcode submit region begin(Prohibit modification and deletion)
    class Solution {
        public int findLength(int[] nums1, int[] nums2) {
            int[] dp = new int[nums1.length + 1];
            int max = 0;
            for (int i = 1; i <= nums1.length; i++) {
                for (int j = nums2.length; j >= 1; j--) {
                    if (nums1[i - 1] == nums2[j - 1]) {
                        dp[j] = dp[j - 1] + 1;
                        max = Math.max(max, dp[j]);
                    } else {
                        dp[j] = 0;
                    }
                }
            }
            return max;
        }
    }
//leetcode submit region end(Prohibit modification and deletion)

}

File: python/leetcode/editor/cn/P718_MaximumLengthOfRepeatedSubarray.py
from typing import List


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def findLength(self, nums1: List[int], nums2: List[int]) -> int:
        n1 = len(nums1)
        n2 = len(nums2)
        if n1 > n2:
            return self.findLength(nums2, nums1)

        b = 131313
        MOD = 10 ** 9 + 7
        N = 1010
        h1 = [0] * N
        h2 = [0] * N
        p = [0] * N
        p[0] = 1

        for i in range(1, n1 + 1):
            h1[i] = (h1[i - 1] * b + nums1[i - 1]) % MOD
            p[i] = (p[i - 1] * b) % MOD
        for i in range(1, n2 + 1):
            h2[i] = (h2[i - 1] * b + nums2[i - 1]) % MOD
            p[i] = (p[i - 1] * b) % MOD

        def func(window_size: int) -> bool:
            i = 1
            s = set()
            while i + window_size - 1 <= n1:
                j = i + window_size - 1
                _hash1 = (h1[j] - h1[i - 1] * p[j - i + 1]) % MOD
                s.add(_hash1)
                i += 1
            i = 1
            while i + window_size - 1 <= n2:
                j = i + window_size - 1
                _hash2 = (h2[j] - h2[i - 1] * p[j - i + 1]) % MOD
                if _hash2 in s:
                    return True
                i += 1
            return False

        left, right = 0, n1
        while left <= right:
            mid = (left + right) >> 1
            if func(mid):
                left = mid + 1
            else:
                right = mid - 1
        return right


# leetcode submit region end(Prohibit modification and deletion)
print(Solution().findLength([1, 2, 3, 2, 1], [3, 2, 1, 4, 7]))
print(Solution().findLength([0, 0, 0, 0, 0], [0, 0, 0, 0, 0]))
